server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -516,28 +862,35 @@ public static IndicesOptions fromXContent(XContentParser parser, @Nullable Indic
             }
         }
 
-        if (wildcardStates == null) {
+        if (wildcardsBuilder == null) {
             throw new ElasticsearchParseException("indices options xcontent did not contain " + EXPAND_WILDCARDS_FIELD.getPreferredName());
+        } else {
+            if (allowNoIndices == null) {
+                throw new ElasticsearchParseException(
+                    "indices options xcontent did not contain " + ALLOW_NO_INDICES_FIELD.getPreferredName()
+                );
+            } else {
+                wildcardsBuilder.allowEmptyExpressions(allowNoIndices);
+            }
         }
         if (ignoreUnavailable == null) {
             throw new ElasticsearchParseException(
                 "indices options xcontent did not contain " + IGNORE_UNAVAILABLE_FIELD.getPreferredName()
             );
         }
-        if (allowNoIndices == null) {
-            throw new ElasticsearchParseException("indices options xcontent did not contain " + ALLOW_NO_INDICES_FIELD.getPreferredName());
-        }
 
+        WildcardOptions wildcardOptions = wildcardsBuilder.build();
+        GeneralOptions generalOptions = generalBuilder.build();
         return IndicesOptions.fromOptions(
```

Comment:
Since we have the new option classes should we just create the indices options using them instead of decomposing and recomposing them?

Reply:
Definitely :D. It was a gradual change, so I completely missed that this step was ready to be made .